crates/core/src/nats/client.rs

```diff
@@ -0,0 +1,262 @@
+use std::{sync::Arc, time::Duration};
+
+use async_nats::connection;
+use bytes::Bytes;
+use tracing::info;
+
+use super::{
+    types::{
+        JetStreamConfig,
+        JetStreamContext,
+        NatsConsumer,
+        NatsStream,
+        PullConsumerConfig,
+    },
+    NatsError,
+    Subject,
+};
+use crate::types::BoxedResult;
+
+#[derive(Debug, Clone)]
+pub struct NatsClient {
+    pub url: String,
+    pub conn_id: String,
+    pub conn: Arc<async_nats::Client>,
+    pub jetstream: Arc<JetStreamContext>,
+}
+
+impl NatsClient {
+    pub async fn connect(
+        url: &str,
+        conn_id: &str,
+        nkey: Option<String>,
+    ) -> Result<Self, NatsError> {
+        let conn_id = conn_id.to_string();
+        let conn = Arc::new(Self::create_conn(url, nkey).await?);
+        let context =
+            Arc::new(async_nats::jetstream::new(conn.as_ref().clone()));
+
+        info!("Connected to NATS server at {}", url);
+        Ok(Self {
+            url: url.to_string(),
+            conn_id,
+            conn,
+            jetstream: context,
+        })
+    }
+
+    pub async fn validate_payload(
+        &self,
+        payload: Bytes,
+        subject_name: &str,
+    ) -> Result<(), NatsError> {
+        let payload_size = payload.len();
+        let max_payload_size = self.conn.server_info().max_payload;
+        if payload_size > max_payload_size {
+            return Err(NatsError::PayloadTooLarge {
+                subject: subject_name.to_string(),
+                payload_size,
+                max_payload_size,
+            });
+        }
+
+        Ok(())
+    }
+
+    pub async fn create_conn(
+        url: &str,
+        nkey: Option<String>,
+    ) -> Result<async_nats::Client, NatsError> {
+        let options = async_nats::ConnectOptions::new()
+            .connection_timeout(Duration::from_secs(30))
+            .max_reconnects(10);
+
+        if let Some(nkey) = nkey {
+            async_nats::connect_with_options(&url, options.nkey(nkey)).await
+        } else {
+            async_nats::connect_with_options(&url, options).await
+        }
+        .map_err(|e| NatsError::ConnectError {
+            url: url.to_owned(),
+            source: e,
+        })
+    }
+
+    pub async fn is_connected(&self) -> Result<bool, NatsError> {
+        let conn_state = self.conn.connection_state();
+        match conn_state {
+            connection::State::Pending => {
+                Err(NatsError::ConnectionPending(self.url.to_owned()))
+            }
+            connection::State::Disconnected => {
+                Err(NatsError::ConnectionDisconnected(self.url.to_owned()))
+            }
+            connection::State::Connected => Ok(true),
+        }
+    }
```

Comment:
With the connection strategy you've written above, and our pod restart policy, we can do away with this to reduce complexity.

Reply:
✅